### ISSUE TITLE
explicit_counter_loop lint suggests invalid fix due to lack of `core::iter::Step` implementation.

### ISSUE THREAD

**Issue by paulmenage** [rust-lang/rust-clippy#16640]
### Summary

The `explicit_counter_loop` lint is suggesting that a loop counter variable and an iterated variable can be combined with `zip()`, but in this case the loop counter type doesn't satisfy `core::iter::Step` so the suggestion doesn't compile.

### Reproducer

Code:

```rust
struct Priority(u8);

impl core::ops::AddAssign<u8> for Priority {
    fn add_assign(&mut self, rhs: u8) {
        self.0 += rhs
    }
}

pub fn foo(x: &[u8]) {
    let mut priority = Priority(1);
    for _val in x {
        priority += 1;
    }
}
```

Current output:

Clippy suggests replacing the loop with `for (priority, _val) in (Priority(1)..).zip(x.iter()) {}`, but this fails to compile:

```
error[E0599]: the method `zip` exists for struct `std::ops::RangeFrom<Priority>`, but its trait bounds were not satisfied
   --> src/lib.rs:17:45
    |
  4 | pub struct Priority(u8);
    | ------------------- doesn't satisfy `Priority: Step`
...
 17 |     for (priority, _val) in (Priority(1)..).zip(x.iter()) {}
    |                                             ^^^ method cannot be called on `std::ops::RangeFrom<Priority>` due to unsatisfied trait bounds
    |
   ::: /Users/paul/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/range.rs:197:1
    |
197 | pub struct RangeFrom<Idx> {
    | ------------------------- doesn't satisfy `std::ops::RangeFrom<Priority>: Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `Priority: Step`
            which is required by `std::ops::RangeFrom<Priority>: Iterator`
            `std::ops::RangeFrom<Priority>: Iterator`
            which is required by `&mut std::ops::RangeFrom<Priority>: Iterator`
note: the trait `Step` must be implemented
   --> /Users/paul/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/iter/range.rs:32:1
    |
 32 | pub trait Step: Clone + PartialOrd + Sized {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```
and `core::iter::Step` is unstable.

Desired output:

Clippy should ignore this snippet.

### Version

```text
rustc 1.95.0-nightly (1ed488274 2026-02-25)
binary: rustc
commit-hash: 1ed488274bec5bf5cfe6bf7a1cc089abcc4ebd68
commit-date: 2026-02-25
host: aarch64-apple-darwin
release: 1.95.0-nightly
LLVM version: 22.1.0

```

### Additional Labels

@rustbot +label l-suggestion-causes-error
